nvflare/private/fed/server/job_runner.py

```diff
@@ -344,67 +345,66 @@ def run(self, fl_ctx: FLContext):
+                                        JobMetaKey.SCHEDULE_COUNT.value: ready_job.meta[
+                                            JobMetaKey.SCHEDULE_COUNT.value
+                                        ],
+                                        JobMetaKey.LAST_SCHEDULE_TIME.value: ready_job.meta[
+                                            JobMetaKey.LAST_SCHEDULE_TIME.value
+                                        ],
+                                        JobMetaKey.SCHEDULE_HISTORY.value: ready_job.meta[
+                                            JobMetaKey.SCHEDULE_HISTORY.value
+                                        ],
+                                    },
+                                    fl_ctx,
                                 )
+
+                            if failed_clients:
+                                deployable_clients = {k: v for k, v in client_sites.items() if k not in failed_clients}
+                            else:
+                                deployable_clients = client_sites
+
+                            self._start_run(
+                                job_id=job_id,
+                                job=ready_job,
+                                client_sites=deployable_clients,
+                                fl_ctx=fl_ctx,
+                            )
+                            with self.lock:
                                 self.running_jobs[job_id] = ready_job
-                                job_manager.set_status(ready_job.job_id, RunStatus.RUNNING, fl_ctx)
-                            except BaseException as e:
-                                if job_id:
-                                    if job_id in self.running_jobs:
-                                        del self.running_jobs[job_id]
-                                    self._stop_run(job_id, fl_ctx)
-                                job_manager.set_status(ready_job.job_id, RunStatus.FAILED_TO_RUN, fl_ctx)
-
-                                deploy_detail = fl_ctx.get_prop(FLContextKey.JOB_DEPLOY_DETAIL)
-                                if deploy_detail:
-                                    job_manager.update_meta(
-                                        ready_job.job_id, {JobMetaKey.JOB_DEPLOY_DETAIL.value: deploy_detail}, fl_ctx
-                                    )
-
-                                self.fire_event(EventType.JOB_ABORTED, fl_ctx)
-                                self.log_error(
-                                    fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {secure_format_exception(e)}"
+                            job_manager.set_status(ready_job.job_id, RunStatus.RUNNING, fl_ctx)
+                        except BaseException as e:
+                            if job_id:
+                                if job_id in self.running_jobs:
+                                    del self.running_jobs[job_id]
+                                self._stop_run(job_id, fl_ctx)
```

Comment:
need to make sure this _stop_run if not put inside a lock, will still function correctly when multiple threads arrive here

Reply:
_stop_run only need to send the stop message / command to the child process. no need to be placed in the lock block.